cosmic-settings/src/pages/power/backend/mod.rs

```diff
@@ -226,3 +227,130 @@ async fn get_power_profiles_proxy<'a>() -> Result<ppdaemon::PowerProfilesProxy<'
         }
     }
 }
+
+#[derive(Default, Debug, Clone)]
+pub struct Battery {
+    pub icon_name: String,
+    pub is_present: bool,
+    pub percent: f64,
+    pub on_battery: bool,
+    pub remaining_duration: Duration,
+    pub remaining_time: String,
+}
+
+async fn get_device_proxy<'a>() -> Result<upower_dbus::DeviceProxy<'a>, zbus::Error> {
+    let connection = match Connection::system().await {
+        Ok(c) => c,
+        Err(e) => {
+            tracing::error!("zbus connection failed. {e}");
+            return Err(e);
+        }
+    };
+
+    match upower_dbus::UPowerProxy::new(&connection).await {
+        Ok(p) => p.get_display_device().await,
+        Err(e) => Err(e),
+    }
+}
+
+async fn get_on_battery_status() -> Result<bool, zbus::Error> {
+    let connection = match Connection::system().await {
+        Ok(c) => c,
+        Err(e) => {
+            tracing::error!("zbus connection failed. {e}");
+            return Err(e);
+        }
+    };
+
+    match upower_dbus::UPowerProxy::new(&connection).await {
+        Ok(p) => p.on_battery().await,
+        Err(e) => Err(e),
+    }
+}
+
+impl Battery {
+    pub async fn update_battery() -> Self {
+        let proxy = get_device_proxy().await;
+
+        if let Ok(proxy) = proxy {
+            let mut is_present: bool = false;
+            let mut percent: f64 = 0.0;
+            let mut on_battery: bool = false;
+            let mut remaining_duration: Duration = Duration::default();
+
+            if let Ok(check) = proxy.is_present().await {
+                is_present = check;
+            }
+
+            if let Ok(percentage) = proxy.percentage().await {
+                percent = percentage.clamp(0.0, 100.0);
+            }
+
+            if let Ok(state) = get_on_battery_status().await {
+                on_battery = state
+            }
```

Comment:
Since this is an async method, you could schedule these to run concurrently like so:

```rs
let (is_present_res, percent_res, on_battery_res) = futures::join!(
    proxy.is_present().map(Result::unwrap_or_default),
    proxy.percentage().map(Result::unwrap_or_default),
    get_on_battery_status().map(Result::unwrap_or_default)
);
```

Reply:
Well, that's so elegant solution and probably more performant. Will add it in a moment :) Thank you! And thanks for pointing those things, I'm just still learning new stuff ^^

Reply:
> Since this is an async method, you could schedule these to run concurrently like so:
> 
> ```rust
> let (is_present, percent, on_battery) = futures::join!(
>     proxy.is_present().map(Result::unwrap_or_default),
>     proxy.percentage().map(Result::unwrap_or_default),
>     get_on_battery_status().map(Result::unwrap_or_default)
> );
> ```

Done!